server.go

```diff
@@ -36,28 +36,50 @@ func (server *Server) Listen(ctx context.Context, address string) (err error) {
 	if err != nil {
 		return
 	}
-	go func() {
-		for {
-			if server.ctx.Err() != nil {
-				close(server.channel)
-				break
+
+	for {
+		done := make(chan error, 1)
+		stop := make(chan struct{})
+
+		if server.ctx.Err() != nil {
+			close(server.channel)
+			break
+		}
+		if conn, e := server.Accept(); e != nil {
+			server.Error = e
+			close(server.channel) // channel cancel ctx
+			for _, c := range server.cancels {
+				c() // all canceled
 			}
-			if conn, e := server.Accept(); e != nil {
-				server.Error = e
-				close(server.channel) // channel cancel ctx 
-				for _, c := range server.cancels{
-					c() // all canceled
+			break
+		} else {
+			// create and call the user callback
+			c, cancel := context.WithCancel(ctx)
+			server.cancels = append(server.cancels, cancel)
```

Comment:
Maybe we coule use a ConnectionManager to reslove this append behaviors enclose in same struct